tests/densify/test_mvs_patchmatchnet.py

```diff
@@ -0,0 +1,120 @@
+DEFAULT_NUM_TRACKS = 30
+DEFAULT_TRACK_POINTS = [Point3(0, 0.5 * float(i), float(i)) for i in range(DEFAULT_NUM_TRACKS)]
+
+# set default camera intrinsics
+DEFAULT_CAMERA_INTRINSICS = Cal3_S2(
+    fx=100.0,
+    fy=100.0,
+    s=0.0,
+    u0=DEFAULT_IMAGE_W // 2,
+    v0=DEFAULT_IMAGE_H // 2,
+)
+# set default camera poses as described in GTSAM example
+DEFAULT_CAMERA_POSES = SFMdata.createPoses(DEFAULT_CAMERA_INTRINSICS)
+# set default camera instances
+DEFAULT_CAMERAS = [
+    PinholeCameraCal3_S2(DEFAULT_CAMERA_POSES[i], DEFAULT_CAMERA_INTRINSICS) for i in range(len(DEFAULT_CAMERA_POSES))
+]
+DEFAULT_NUM_CAMERAS = len(DEFAULT_CAMERAS)
+# the number of valid images should be equal to the number of cameras (with estimated pose)
+DEFAULT_NUM_IMAGES = DEFAULT_NUM_CAMERAS
+
+# build dummy image dictionary with default image shape
+DEFAULT_DUMMY_IMAGE_DICT = {
+    i: Image(value_array=np.zeros([DEFAULT_IMAGE_H, DEFAULT_IMAGE_W, DEFAULT_IMAGE_C], dtype=int))
+    for i in range(DEFAULT_NUM_IMAGES)
+}
+
+
+class TestMVSPatchmatchNet(unittest.TestCase):
+    """Unit tests for PatchmatchNet method."""
+
+    def setUp(self) -> None:
+        """Set up the image dictionary and gtsfm result for the test."""
+        super().setUp()
+
+        # set the number of images as the default number
+        self._num_images = DEFAULT_NUM_IMAGES
+
+        # set up the default image dictionary
+        self._img_dict = DEFAULT_DUMMY_IMAGE_DICT
+
+        # initialize sfm result
+        self._sfm_result = self.get_dummy_gtsfm_data()
+
+        # use patchmatchnet to recontruct dense point cloud
+        self._dense_points = MVSPatchmatchNet().densify(
+            self._img_dict, self._sfm_result, max_num_views=DEFAULT_NUM_IMAGES
+        )
+
+    def get_dummy_gtsfm_data(self) -> GtsfmData:
+        """Create a new GtsfmData instance, add dummy cameras and tracks, and draw the track points on all images"""
+        sfm_result = GtsfmData(self._num_images)
+        # Assume all default cameras are valid cameras, add toy data for cameras
+        for i in range(DEFAULT_NUM_CAMERAS):
+            sfm_result.add_camera(i, DEFAULT_CAMERAS[i])
+
+        # Calculate the measurements under each camera for all track points, then add toy data for tracks:
+        for j in range(DEFAULT_NUM_TRACKS):
+            world_x = DEFAULT_TRACK_POINTS[j]
+            track_to_add = SfmTrack(world_x)
+
+            for i in range(DEFAULT_NUM_CAMERAS):
+                u, v = sfm_result.get_camera(i).project(world_x)
+
+                # color the track point with (r, g, b) = (255, 255, 255)
+                self._img_dict[i].value_array[int(v + 0.5), int(u + 0.5), :] = 255
+
+                track_to_add.add_measurement(idx=i, m=[u, v])
+            sfm_result.add_track(track_to_add)
+
+        return sfm_result
+
+    def test_patchmatchnet_result(self) -> None:
+        """Test whether the result point cloud calculated by PatchmatchNet correctly reflects the original shape"""
```

Comment:
could you share an image of the result depth maps and point clouds to explain this a bit more?

Reply:
Designed model for unit test
![dummy_gt](https://user-images.githubusercontent.com/38122010/135887498-777d3f37-ef98-4b13-9b2c-a69e63c8bee5.png)

Generated images from 8 default cameras
![gen_images](https://user-images.githubusercontent.com/38122010/135887512-2f286704-d5a6-4072-ac3d-9626e7e1d64c.png)

Output point cloud
![1](https://user-images.githubusercontent.com/38122010/135887529-fa7b110e-ec8a-4e5c-897b-31830f9b536d.png)

Reply:
do we need to output depth maps?

Reply:
I think the number of output points cannot be precisely estimated